File: 289__Game_of_Life.py
class Solution(object):
    def gameOfLife(self, board):
        """
        :type board: List[List[int]]
        :rtype: void Do not return anything, modify board in-place instead.
        """
        len_m = len(board)
        len_n = len(board[0])

        for m in xrange(len_m):
            for n in xrange(len_n):
                live_neighbor = 0
                for temp_m in xrange(max(0, m - 1), min(len_m - 1, m + 1) + 1):
                    for temp_n in xrange(max(0, n - 1), min(len_n - 1, n + 1) + 1):
                        if (m, n) != (temp_m, temp_n) and (board[temp_m][temp_n] == 1 or board[temp_m][temp_n] == 2):
                            live_neighbor += 1
                if board[m][n] == 1:
                    if live_neighbor < 2 or live_neighbor > 3:
                        board[m][n] = 2
                else:
                    if live_neighbor == 3:
                        board[m][n] = 3

        for m in xrange(len_m):
            for n in xrange(len_n):
                if board[m][n] == 2:
                    board[m][n] = 0
                if board[m][n] == 3:
                    board[m][n] = 1

                    

File: Java/289__Game_of_Life.java
class Solution {
    public void gameOfLife(int[][] board) {
        if (board.length == 0 || board[0].length == 0) {
            return;
        }
        int[] dx = {-1, 0, 1, -1, 1, -1, 0, 1};
        int[] dy = {-1, -1, -1, 0, 0, 1, 1, 1};
        for (int i = 0; i < board.length; i++) {
            for (int j = 0; j < board[0].length; j++) {
                int count = 0;
                for (int k = 0; k < 8; k++) {
                    if (0 <= i + dx[k] && i + dx[k] < board.length && 0 <= j + dy[k] && j + dy[k] < board[0].length) {
                        if (board[i + dx[k]][j + dy[k]] % 2 == 1) {
                            count++;
                        }
                    }
                }
                if (board[i][j] == 1 && !(count == 2 || count == 3)) {
                    board[i][j] = 3;
                }
                if (board[i][j] == 0 && count == 3) {
                    board[i][j] = 2;
                }
            }
        }
        for (int i = 0; i < board.length; i++) {
            for (int j = 0; j < board[0].length; j++) {
                if (board[i][j] == 3) {
                    board[i][j] = 0;
                }
                if (board[i][j] == 2) {
                    board[i][j] = 1;
                }
            }
        }
    }

}
